pkg/kubelet/eviction/helpers.go

```diff
@@ -162,3 +183,360 @@ func parseGracePeriods(expr string) (map[Signal]time.Duration, error) {
+func (ms *multiSorter) Less(i, j int) bool {
+	p1, p2 := ms.pods[i], ms.pods[j]
+	var k int
+	for k = 0; k < len(ms.cmp)-1; k++ {
+		cmpResult := ms.cmp[k](p1, p2)
+		// p1 is less than p2
+		if cmpResult < 0 {
+			return true
+		}
+		// p1 is greater than p2
+		if cmpResult > 0 {
+			return false
+		}
+		// we don't know yet
+	}
+	// the last cmp func is the final decider
+	return ms.cmp[k](p1, p2) < 0
+}
+
+// qos compares pods by QoS (BestEffort < Burstable < Guaranteed)
+func qos(p1, p2 *api.Pod) int {
+	qosP1 := qosutil.GetPodQos(p1)
+	qosP2 := qosutil.GetPodQos(p2)
+	// its a tie
+	if qosP1 == qosP2 {
+		return 0
+	}
+	// if p1 is best effort, we know p2 is burstable or guaranteed
+	if qosP1 == qosutil.BestEffort {
+		return -1
+	}
+	// we know p1 and p2 are not besteffort, so if p1 is burstable, p2 must be guaranteed
+	if qosP1 == qosutil.Burstable {
+		if qosP2 == qosutil.Guaranteed {
+			return -1
+		}
+		return 1
+	}
+	// ok, p1 must be guaranteed.
+	return 1
+}
+
+// memory compares pods by largest consumer of memory relative to request.
+func memory(stats statsFunc) cmpFunc {
+	return func(p1, p2 *api.Pod) int {
+		p1Stats, found := stats(p1)
+		// if we have no usage stats for p1, we want p2 first
+		if !found {
+			return -1
+		}
+		// if we have no usage stats for p2, but p1 has usage, we want p1 first.
+		p2Stats, found := stats(p2)
+		if !found {
+			return 1
+		}
+		// if we cant get usage for p1 measured, we want p2 first
+		p1Usage, err := podUsage(p1Stats)
+		if err != nil {
+			return -1
+		}
+		// if we cant get usage for p2 measured, we want p1 first
+		p2Usage, err := podUsage(p2Stats)
+		if err != nil {
+			return 1
+		}
+
+		// adjust p1, p2 usage relative to the request (if any)
+		p1Memory := p1Usage[api.ResourceMemory]
+		p1Spec := core.PodUsageFunc(p1)
+		p1Request := p1Spec[api.ResourceRequestsMemory]
+		p1Memory.Sub(p1Request)
+
+		p2Memory := p2Usage[api.ResourceMemory]
+		p2Spec := core.PodUsageFunc(p2)
+		p2Request := p2Spec[api.ResourceRequestsMemory]
+		p2Memory.Sub(p2Request)
+
+		// if p2 is using more than p1, we want p2 first
+		return p2Memory.Cmp(p1Memory)
+	}
+}
+
+// disk compares pods by largest consumer of disk relative to request.
+func disk(stats statsFunc) cmpFunc {
+	return func(p1, p2 *api.Pod) int {
+		p1Stats, found := stats(p1)
+		// if we have no usage stats for p1, we want p2 first
+		if !found {
+			return -1
+		}
+		// if we have no usage stats for p2, but p1 has usage, we want p1 first.
+		p2Stats, found := stats(p2)
+		if !found {
+			return 1
+		}
+		// if we cant get usage for p1 measured, we want p2 first
+		p1Usage, err := podUsage(p1Stats)
+		if err != nil {
+			return -1
+		}
+		// if we cant get usage for p2 measured, we want p1 first
+		p2Usage, err := podUsage(p2Stats)
+		if err != nil {
+			return 1
+		}
+
+		// disk is best effort, so we don't measure relative to a request.
+		// TODO: add disk as a guaranteed resource
+		p1Disk := p1Usage[api.ResourceStorage]
```

Comment:
Resource storage is pretty generic and weirdly it only applies to volumes as per API comments. Should `local-disk` be a first class resource?